### PR TITLE
CPM-728: Create IdentifierGenerator SqlRepository

### PR DESCRIPTION
**Pull request by antoinearrive**
**Description (for Contributor and Core Developer)**

Added SQL implementation without validations

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    private Connection $connection;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->connection = $this->get('database_connection');
+    }
+
+    /** @test */
+    public function it_adds_indentifier_generator_table_if_not_present()
```
Comment (mmetayer) [inline upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php:24]:
```suggestion
    public function it_creates_the_indentifier_generator_table_if_not_present(): void
```

File: upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        $this->reExecuteMigration(self::MIGRATION_LABEL);
```
Comment (mmetayer) [inline upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php:26]:
```suggestion
        $this->connection->executeStatement('DROP TABLE IF EXISTS pim_catalog_identifier_generator');
        Assert::assertFalse($this->tableExists());
        $this->reExecuteMigration(self::MIGRATION_LABEL);
```

File: upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        $this->assertTrue($this->tableExists());
```
Comment (mmetayer) [inline upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php:27]:
```suggestion
        Assert::assertTrue($this->tableExists());
        $this->reExecuteMigration(self::MIGRATION_LABEL);
        Assert::assertTrue($this->tableExists());
        
```

Comment (mmetayer) [reply upgrades/test_schema/Version_7_0_20220927080024_add_identifier_generator_table_Integration.php:27]: this way you ensure that the migration won't fail if the migration is executed twice for any reason

File: components/identifier-generator/back/tests/.php_cd.php [hunk 4]
```diff
@@ -32,12 +32,18 @@
     $builder->only(
         [
             'Akeneo\Pim\Automation\IdentifierGenerator\Domain',
+            'Akeneo\Platform\Bundle\InstallerBundle\Event\InstallerEvents',
 
+            'Ramsey\Uuid\Uuid',
+            'Doctrine\DBAL\Exception',
+            'Doctrine\DBAL\Connection',
+            'Doctrine\DBAL\DBALException',
```
Comment (mmetayer) [inline components/identifier-generator/back/tests/.php_cd.php:40]:
```suggestion
```

File: upgrades/schema/Version_7_0_20220927080024_add_identifier_generator_table.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        $this->addSql(<<<SQL
+            CREATE TABLE IF NOT EXISTS pim_catalog_identifier_generator (
+                `uuid` binary(16) PRIMARY KEY,
+                `code` VARCHAR(100) NOT NULL,
+                `conditions` JSON NOT NULL DEFAULT ('{}'),
+                `structure` JSON NOT NULL DEFAULT ('{}'),
+                `labels` JSON NOT NULL DEFAULT ('{}') ,
+                `target` VARCHAR(100) NOT NULL,
+                `delimiter` VARCHAR(100),
+                INDEX code_identifier_generator (code)
```
Comment (mmetayer) [inline upgrades/schema/Version_7_0_20220927080024_add_identifier_generator_table.php:21]:
```suggestion
                UNIQUE INDEX unique_identifier_generator_code (code)
```

File: components/identifier-generator/back/tests/Integration/Repository/SqlIdentifierGeneratorRepositoryIntegration.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            Conditions::fromNormalized([]),
+            Structure::fromArray([FreeText::fromString('abc')]),
+            LabelCollection::fromNormalized(['fr' => 'Générateur']),
+            Target::fromString('sku'),
+            Delimiter::fromString('-'),
+        );
+
+        $this->identifierGeneratorRepository->save($identifierGenerator);
+    }
+
```
Comment (mmetayer) [inline components/identifier-generator/back/tests/Integration/Repository/SqlIdentifierGeneratorRepositoryIntegration.php:65]: can you add a test which tries to insert 2 rows with trhe same code and check that it throws an UnableToSaveException please?

File: components/identifier-generator/back/src/Domain/Model/Property/FreeText.php [hunk 7]
```diff
@@ -18,6 +18,24 @@ private function __construct(
...
+    {
+        return 'free_text';
+    }
+
+    public static function fromNormalized(array $fromNormalized): PropertyInterface
+    {
+        return self::fromString($fromNormalized['string']);
+    }
+
+    public function normalize()
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Domain/Model/Property/FreeText.php:31]:
```suggestion
    public function normalize(): array
```

File: components/identifier-generator/back/src/Infrastructure/Repository/SqlIdentifierGeneratorRepository.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            throw new UnableToSaveIdentifierGeneratorException(sprintf('Cannot save the identifier generator "%s"', $identifierGenerator->code()->asString()));
+        }
+    }
+
+    /**
+     * @inerhitDoc
+     */
+    public function get(string $identifierGeneratorCode): ?IdentifierGenerator
+    {
+        if (trim($identifierGeneratorCode) == '') {
```
Comment (Beninho) [inline components/identifier-generator/back/src/Infrastructure/Repository/SqlIdentifierGeneratorRepository.php:63]: prefer strict comparison

Comment (antoinearrive) [reply components/identifier-generator/back/src/Infrastructure/Repository/SqlIdentifierGeneratorRepository.php]: done

File: components/identifier-generator/back/src/Domain/Model/Condition/Conditions.php [hunk 9]
```diff
@@ -21,13 +21,15 @@ private function __construct(
...
-     * @param ConditionInterface[] $conditions
-     */
     public static function fromArray(array $conditions): self
     {
         Assert::allIsInstanceOf($conditions, ConditionInterface::class);
 
         return new self($conditions);
     }
+
+    public function normalize(): array
```
Comment (cemorin) [inline components/identifier-generator/back/src/Domain/Model/Condition/Conditions.php:31]: Why is it added (just to understand)?

**Review by cemorin — APPROVED** [review]

**Review by Beninho — APPROVED** [review]

File: components/identifier-generator/back/src/Domain/Model/Condition/Conditions.php (hunk 9 above)
Comment (antoinearrive) [reply components/identifier-generator/back/src/Domain/Model/Condition/Conditions.php:31]: I added it to be able to use it in the sql record and to harmonize with the other properties that call "normalize"

**Pull request merged**